### ISSUE TITLE
Users are logged out after SDK upgrade

### ISSUE THREAD

**Issue by PLR2388** [parse-community/Parse-SDK-Android#1158]
### New Issue Checklist

- [X] I am not disclosing a [vulnerability](https://github.com/parse-community/Parse-SDK-Android/security/policy).
- [X] I am not just asking a [question](https://github.com/parse-community/.github/blob/main/SUPPORT.md).
- [X] I have searched through [existing issues](https://github.com/parse-community/Parse-SDK-Android/issues?q=is%3Aissue).
- [X] I can reproduce the issue with the latest version of [Parse Server](https://github.com/parse-community/parse-server/releases) and the [Parse Android SDK](https://github.com/parse-community/Parse-SDK-Android/releases). 

### Issue Description

ParseUser.getCurrentUser() becomes null after an update from `1.25.0` to `3.0.0`. 

### Steps to reproduce

- Log in to an app with `1.25.0`
- Update your app with parse `3.0.0`
- See that you're logout

### Actual Outcome

I'm logout

### Expected Outcome

I should be login as before

### Environment

Parse Android SDK
- SDK version: `3.0.0`
- Operating system version: `API 30`

Server
- Parse Server version: `4.10.4` (I tried in a staging environment with `5.2.0` too)
- Operating system: `Ubuntu 20.04.2 LTS`
- Local or remote host (AWS, Azure, Google Cloud, Heroku, Digital Ocean, etc): `Remote host`

Database
- System (MongoDB or Postgres): `MongoDB`
- Database version: `4.4.13`
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): `MongoDB Atlas`

**Comment by qmetzler-luna** [comment]
After some more testing, the issue seems to happen when updating the lib from [2.1.0](https://github.com/parse-community/Parse-SDK-Android/releases/tag/2.1.0) to [3.0.0](https://github.com/parse-community/Parse-SDK-Android/releases/tag/3.0.0)

**Comment by mtrezza** [comment]
If you look at the commits, are you able to tell which commit may cause this?

**Comment by PLR2388** [comment]
I think the commit that may cause this is this [one](https://github.com/parse-community/Parse-SDK-Android/commit/6fad701158950168fcfcb8a3730ae844513e99d4) as the others before [3.0.0](https://github.com/parse-community/Parse-SDK-Android/releases/tag/3.0.0) are just renaming or adding documentation 🤔

**Comment by L3K0V** [comment]
In version [3.0.0](https://github.com/parse-community/Parse-SDK-Android/releases/tag/3.0.0) a breaking changes was introduced:

> The following deprecated methods are removed: Parse.getParseDir() (use getParseCacheDir(String) or getParseFilesDir(String) instead)

Is very likely that internally the user was saved in the old place and after the upgrade the new one is used where the user cannot be found. If that is correct we can think about implementing a proper migration of the user or/and installation. From what I remember, there was actually some kind of migration, but I need to double-check what covers.

Nice catch @qmetzler-luna

**Comment by mtrezza** [comment]
If that is really the case, we may add this to the changelog even before we publish a fix; this could be quite a severe issue if all users who update their app get logged out.

**Comment by mtrezza** [comment]
Sounds like a rather trivial fix to me. Would you want to open a PR for this?

**Comment by mman** [comment]
@mtrezza What's your idea of a trivial fix?

If we just revert back to loading at least `currentUser`, `currentInstallation`, and `installationId` from old location, then all apps out in the wild on 3.0.0 - that were forced to re-login will break as they now contain old data in old location and new data in new location. The old data are potentially/very probably not in sync. And I believe that there was some hidden motivation behind shuffling the files around in the app data directory. Right?

So the fix should be more elaborate, essentially trying to see if the file is present at new location, and use that (assuming login happened on 3.0.0), and if it is not present, trying to move previous file from old location to a new location and use that (assuming user was logged in on 1.x or 2.x).

In any case, looking that 3.0.0 was released Nov 21, it's potentially now out in the wild for several months ideal fix would not break anything again.

Your thoughts?
Martin

**Comment by mman** [comment]
@mtrezza More over, `currentUser` is being now loaded from `Parse.getParseFilesDir`, whereas `currentConfig`, `currentInstallation`, and `installationId` are loaded from `ParsePlugins.getFilesDir`. They all if I read the code correctly end up in the same location in the end, but any reason why these are not aligned?

thanks,
Martin

**Comment by mtrezza** [comment]
First of all, I'm glad we identified this issue. It actually sounds more like a bug than a breaking change, because there is virtually nothing the developer can do except for writing their own migration routine, which IMO should be done by the SDK.

A trivial fix could be:
```
if newPath doesn't exist {
  move all content from old path to new path
}

delete old path
```

If someone has both paths already then I wouldn't do any migration and just delete the old path.

**Comment by L3K0V** [comment]
I will have no availability this and next 2-3 weeks. Sorry. Only have time to follow discussions and doing some reviews if necessary.

**Comment by mtrezza** [comment]
@L3K0V If you could do a review that would already help a lot. @mman Would you be up to write a fix? We'll a bounty to this for urgency so it's compensated as well.

**Comment by mtrezza** [comment]
Thanks to @rommansabbir who submitted the PR for cache migration to fix this issue. Could anyone try this out to verify that it works?

**Comment by L3K0V** [comment]
https://github.com/parse-community/Parse-SDK-Android/commit/7d0faa38bf3ccc85b715dc61965d5b82103c19a4#diff-6cb14853da236c30db3f14e4005ea40c5c45cb8fcb9fc49541693186620de1dcR60

@mtrezza I remember that there was some migration functionality, but for the push notifications. I started wondering if this should be reverted? But I assumed that who is using the 1.26 and early versions which was introduced for a long long time are already migrated. For me this was a very old legacy migration.

Still, similar but very slim and simplified version of such migration should be used to migrate the user, installation etc. I already posted into the PR

**Comment by mtrezza** [comment]
Yes, I would remove old migration routines only after a *very* long time, if they don't hurt being there.

**Comment by mtrezza** [comment]
Note: conceptual conversation about migration is happening in https://github.com/parse-community/Parse-SDK-Android/pull/1168

**Comment by PLR2388** [comment]
Btw thank you all :)

**Comment by mtrezza** [comment]
Thank you for kicking it off by opening the issue ;-)
